### PR TITLE
Generate girder job from celery task API

### PR DESCRIPTION
**Pull request by kotfic**
This PR intends to provide girder rest endpoints the option of calling girder_worker tasks via the [celery task api](http://docs.celeryproject.org/en/latest/userguide/calling.html#guide-calling).  As an example:

```python
from girder.api import access
from girder.api.describe import Description, describeRoute
from girder.api.rest import Resource, filtermodel
from gwexample.analyses.tasks import fibonacci

class Fib(Resource):
    def __init__(self):
        super(Fib, self).__init__()
        self.resourceName = 'fib'

        self.route('POST', ('celery', ), self.run_fib)

    @access.token
    @filtermodel(model='job', plugin='jobs')
    @describeRoute(
        Description('Fire off a a fibonacci job using fibonacci.delay')
        .param('number', 'The upper bound of the fibonacci sequence',
               dataType='integer'))
    def run_fib(self, params):
        number = int(params.get('number', 10))

        a = fibonacci.delay(number, girder_user=self.getCurrentUser())

        return a.job

```

This is accomplished by using a custom celery task class (here ```GWTask```) and overriding the apply_async method to generate the girder job model.   Note that girder specific imports are guarded with-in the task to prevent a hard dependency on girder. ```apply_async(...)``` is called producer side within the context of a girder REST request, unless it is called outside of girder, in which case the imports will fail and the ```GWTask``` class will pass all arguments through to ```celery.Task.apply_async(...)```.

### CODE REVIEW

**Comment by kotfic** [comment]
@cjh1 @zachmullen @danlamanna I will clean up styles/tests once i have ticked off a few other things i want to accomplish with this PR,  I just wanted to get this in front of some people before I got too far ahead of myself.  PTAL!

File: girder_worker/app.py [hunk 1]
```diff
@@ -1,12 +1,90 @@
...
 import traceback as tb
-from celery import Celery, __version__
+from celery import Celery, Task, __version__
 from distutils.version import LooseVersion
 from celery.signals import (task_prerun, task_postrun,
                             task_failure, task_success, worker_ready)
 from .utils import JobStatus
 
 
+class GWTask(Task):
```
Comment (cjh1) [inline girder_worker/app.py]: Do we really need the GW prefix, we are already in the girder_worker package?

**Comment by cjh1** [comment]
@kotfic Are you planning on refactoring this so the job model is created in the ```before_task_publish``` signal  handler? We should also add a test?

**Comment by kotfic** [comment]
@cjh1 tests added - PTAL!

**Comment by cjh1** [comment]
@kotfic Excellent, I will take a look today.

File: girder_worker/app.py [hunk 2]
```diff
@@ -1,12 +1,136 @@
...
+
+            except ImportError:
+                # Make a rest request to get the job info
+                return None
+
+        return self._job
+
+    @job.setter
+    def job(self, value):
+        self._job = value
```
Comment (cjh1) [inline girder_worker/app.py]: It's the job property read-only in terms of the API? I would think it should be assigned via some internal  mechanism such as just assigning to self._job.

File: girder_worker/app.py [hunk 3]
```diff
@@ -1,12 +1,136 @@
...
+    _girder_job_type = None
+    _girder_job_public = False
+    _girder_job_handler = 'celery_handler'
+    _girder_job_other_fields = {}
+
+    special_headers = ['girder_token', 'girder_user']
+
+    def AsyncResult(self, task_id, **kwargs):
+        return GirderAsyncResult(task_id, backend=self.backend,
+                                 task_name=self.name, **kwargs)
```
Comment (cjh1) [inline girder_worker/app.py]: What about app?

File: girder_worker/app.py [hunk 4]
```diff
@@ -1,12 +1,136 @@
...
+            from girder.utility.model_importer import ModelImporter
+            from girder.plugins.worker import utils
+            from girder.api.rest import getCurrentUser
+
+            job_model = ModelImporter.model('job', 'jobs')
+
+            user = headers.pop('girder_user', getCurrentUser())
+            token = headers.pop('girder_token', None)
+
+            task_args, task_kwargs = body[0], body[1]
```
Comment (cjh1) [inline girder_worker/app.py]: Is the structure of body a public interface/stable/defined?

File: girder_worker/utils.py [hunk 5]
```diff
@@ -3,6 +3,25 @@
 import sys
 
 
+def girder_job(title=None, type=None, public=False,
+               handler='celery_handler', otherFields=None):
```
Comment (cjh1) [inline girder_worker/utils.py]: Could we document what these parameters mean?

File: tests/task_signal_test.py [hunk 6]
```diff
@@ -22,6 +40,94 @@ def setUp(self):
...
+
+        with mock_worker_plugin_utils():
+            girder_before_task_publish(**inputs)
+
+        self.assertTrue(not gcu.called)
+        self.assertTrue(not create_job.called)
+
+    @mock.patch('girder.utility.model_importer.ModelImporter')
+    @mock.patch('girder.api.rest.getCurrentUser')
+    def test_girder_before_task_publish_called_with_no_headers_create_job_inputs(self, gcu, mi):
```
Comment (cjh1) [inline tests/task_signal_test.py]: Longest test name in history :smile_cat:

File: girder_worker/app.py [hunk 7]
```diff
@@ -1,12 +1,136 @@
...
+
+    @property
+    def job(self):
+        if self._job is None:
+            try:
+                from girder.utility.model_importer import ModelImporter
+                job_model = ModelImporter.model('job', 'jobs')
+
+                try:
+                    return list(job_model.find({'celeryTaskId': self.task_id}))[0]
```
Comment (danlamanna) [inline girder_worker/app.py]: You should be able to use `findOne` here.

File: girder_worker/app.py [hunk 8]
```diff
@@ -1,12 +1,136 @@
...
+class GirderAsyncResult(AsyncResult):
+    def __init__(self, *args, **kwargs):
+        self._job = None
+        super(GirderAsyncResult, self).__init__(*args, **kwargs)
+
+    @property
+    def job(self):
+        if self._job is None:
+            try:
+                from girder.utility.model_importer import ModelImporter
```
Comment (danlamanna) [inline girder_worker/app.py]: Could you leave a comment about how this is importing Girder only in the event that these are being called on the Girder side?

File: girder_worker/app.py (hunk 2 above)
Comment (kotfic) [inline girder_worker/app.py]: yeah this is a relic from the apply_async() approach. I will remove it

File: girder_worker/app.py (hunk 4 above)
Comment (kotfic) [inline girder_worker/app.py]: yes (http://docs.celeryproject.org/en/latest/internals/protocol.html#definition)

File: girder_worker/utils.py (hunk 5 above)
Comment (kotfic) [inline girder_worker/utils.py]: Sure,  for reference they are the parameters used by the girder job model

**Comment by cjh1** [comment]
A general comment, we have a mix of snake_case and camelCase for functions names, we should standardize, as girder allows camelCase, we should probably stick to that?

**Comment by jeffbaumes** [comment]
Where possible, we should use or transition to Girder's code conventions, which I think is camelCase, correct?

**Comment by cjh1** [comment]
Correct

**Comment by cjh1** [comment]
Looks like I aready opened a style issue #105

**Pull request merged**